fix(TTML): Correctly handle multiple samples in a segment

julijane opened the pull request:
Fixes https://github.com/shaka-project/shaka-player/issues/8087

Implements handling of multiple samples in a MP4/ISOBMFF/DASH TTML segment/fragment. Such segments are allowed by ISO14496-12 and ISO23000-19. gpac creates such segments. The prior code just treated the full MDAT as one TTML XML document and tried to parse it in whole without accounting for sample(s). A testcase is included which was created by taking the testdata from ttml-segment.mp4 and splitting the subtitles into two independent TTML-XML documents, which then were put as individual samples.

The testdata for the prior existing multiple MDAT testcase was invalid. It was created by taking the same ttml-segment.mp4 as a source and just duplicating the MDAT box, but without then also fixing the TRUN box. The duplicated data was thus not referenced. The test case still worked, because the prior code did not look at the TRUN box and the sample specification at all and just handled any full MDAT box = 1 sample. The testdata was replaced with a new file, which is basically the same as for the multiple samples case, but with the two samples split into two MDAT boxes.

theodab commented on lib/text/mp4_ttml_parser.js:
```diff
@@ -80,67 +82,117 @@ shaka.text.Mp4TtmlParser = class {
-            payload = payload.concat(
-                this.parser_.parseMedia(data, time, uri, /* images= */ []));
-          }
+          // We first only collect all mdat, in case the mp4 has
```
This comment is confusingly worded. I assume you meant something like this?

> We collect all of the mdats first, before parsing any of them. This is necessary in case the mp4 has multiple mdats.

theodab commented on lib/text/mp4_ttml_parser.js:
```diff
@@ -80,67 +82,117 @@ shaka.text.Mp4TtmlParser = class {
+    let defaultSampleSize = null;
+
+    /** @type {!Array<Uint8Array>} */
+    const mdat = [];
```
I would call this variable `mdats`, to make it clearer that it is an array that contains multiple mdat sub-arrays, instead of a single mdat that we are concatenating data on to (like how we do with the `payload`).

theodab commented on lib/text/mp4_ttml_parser.js:
```diff
@@ -80,67 +82,117 @@ shaka.text.Mp4TtmlParser = class {
+        shaka.util.Uint8ArrayUtils.concat(...mdat);
+
+    let sampleOffset = 0;
+    for (let sampleNum=0; sampleNum < sampleSizes.length; sampleNum++) {
```
```suggestion
    for (let sampleNum = 0; sampleNum < sampleSizes.length; sampleNum++) {
```

julijane commented on lib/text/mp4_ttml_parser.js:
```diff
@@ -6,11 +6,13 @@
 goog.require('shaka.util.BufferUtils');
 goog.require('shaka.util.Error');
 goog.require('shaka.util.Mp4Parser');
+goog.require('shaka.util.Mp4BoxParsers');
```
Done. Seems I missed that when I redid the fix (I first did it on a way older shaka player version).

julijane commented:
Oops. When looking on my change again, I noticed I missed to change one line when I moved to a Map for the subSampleSizes as suggested by theodab. Just fixed that. Looks like that does not reset approvals here though?

avelad commented:
> Oops. When looking on my change again, I noticed I missed to change one line when I moved to a Map for the subSampleSizes as suggested by theodab. Just fixed that. Looks like that does not reset approvals here though? But also seems pipelines don't run again now?

This is your first PR, I have to manually approve the tests. Sorry for the inconvenience!

avelad commented:
> > This is your first PR, I have to manually approve the tests. Sorry for the inconvenience!
> 
> No worries, just wanted to point out that I had to make another change to the PR.
> 
> One of the tests (Firefox on Windows) was failing also at the last run, but in WebVTT code, so not related to my change.

Don't worry, it's just a bug that we have to update some screenshots, but that machine in the lab is not working right now, but I'll fix it in a few days.
